[Major] Change static variables for injections

saantiaguilera opened the pull request:
- Change static references across code (which served as injections, eg.
    the TokenHandler) for injections in constructors of each component.
This makes the code more verbose, but allows us to scale further since
we are no longer depending of a static reference, we can simply inject
what we want where we like.
- Create global services, as we want a single instance of each of them
(to avoid spawning workers / warming up / re-creating conections or even
 holding more than the necessaries)

saantiaguilera committed:
Merge branch 'feature/sa/controller_code' into feature/sa/jwt_on_code_controller

saantiaguilera committed:
[C3PO-13] Add JWT implementation as middleware

- Add to controllers the possibility of adding own middlewares
- Add default middlewares to engine (log / recover)
- Create middleware of JWT implementation for code
- Add logic to middleware that validates only on user_id parameter (on a future
    step we will probably want to add more logic, such as allowing
    professors to read students code)

We didn't use a global middleware for authenticating because not all
endpoints will contain authentication AND not necessarily all
authenticated endpoints will be the same (eg. Code will authenticate
    against the student + professor, while users API will probably only
    authenticate against the own user)

saantiaguilera committed:
[C3PO-13] Add code controllers implementation

- Add GET implementation
- Add POST implementation
- Add PUT implementation
- Add golden file retriever
- Add CodeService abstraction (in a future this will be a DB + OS or
    something like that)
- Add unit tests of each file

saantiaguilera committed:
[C3PO-13] Add integration tests

- Add integration tests about code API
- For the moment they aren't completely realistic as we don't have yet a
code service. We will be mocking that (and will be changed when
    developed)

saantiaguilera committed:
[C3PO-13] Change malformed header for JWT

- Refactor 'Authentication' for 'Authorization', as I was probably having a
stroke when I wrote it.

saantiaguilera committed:
[C3PO-13] Remove duplicate middlewares

- Remove duplicate middlewares per warnings outputted
- Add a sleep before making requests to ensure (as much as possible)
  server ignition

saantiaguilera committed:
[C3PO-13] Use dial for checking conn liveliness

- Instead of doing a sleep which makes our tests unstable, we will be
using a TCP dial (with 5 retries) trying to stablish a connection with
our server. If it fails the test will be consider a failure.

saantiaguilera committed:
Merge branch 'master' into feature/sa/jwt_on_code_controller

saantiaguilera committed:
Merge branch 'feature/sa/jwt_on_code_controller' of github.com:arpb2/C-3PO into feature/sa/code_controllers_implementation

saantiaguilera committed:
Merge branch 'feature/sa/code_controllers_implementation' of github.com:arpb2/C-3PO into feature/sa/integration_tests

saantiaguilera committed:
[C3PO-14] Users API

- Add controller for users API providing basic GET, POST, PUT, DELETE
- Add a binder pattern with dispatch for registering controllers (as the
    registration was becoming god)
- Segregate registration of controllers from ServerEngine interface to
tighten app binders visibliity
- Fix wordings of Authentication / Authorization in some messages

saantiaguilera committed:
[C3PO-14] Refactor interface names

saantiaguilera committed:
[C3PO-14] User authentication middleware

- Add a single middleware masked by 2 proxies that enable/disable the
teacher ("admin") check
- Add proxy middlewares
- Consume proxy middlewares from controllers
- Add expected behaviour to the middlewares + proxies

saantiaguilera committed:
[C3PO-14] Add user model

- Add user model
- Add credential model
- Add authenticated user model

saantiaguilera committed:
[C3PO-14] Auth improvements

- Centralize halts in a single command
- Change middleware of authentication into a strategy injector, to avoid
having weird couplings with business rules

saantiaguilera committed:
[C3PO-14] Improvements to auth

- Add improvements to middleware usage, allowing to inject strategies
for richer validations (instead of coupling ourselves to business rules)
- Add a centralized method for halting requests, in case we change in a
future the error response structure

saantiaguilera committed:
[C3PO-14] Refactor token handler exposing

- Expose token handler directly from its consumer, to avoid a weird
dependency graph. The consumer knows his abstraction

saantiaguilera committed:
[Major] Change static references for injections

- Change static references across code (which served as injections, eg.
    the TokenHandler) for injections in constructors of each component.
This makes the code more verbose, but allows us to scale further since
we are no longer depending of a static reference, we can simply inject
what we want where we like.
- Create global services, as we want a single instance of each of them
(to avoid spawning workers / warming up / re-creating conections or even
 holding more than the necessaries)

mastanca commented on src/api/middleware/auth/authentication_handler.go:
```diff
@@ -3,28 +3,25 @@ package auth
 
-var TokenHandler auth.TokenHandler = jwt.TokenHandler{
-	Secret: jwt.FetchJwtSecret(),
+type AuthenticationStrategy interface {
```
👍

mastanca replied:
Te va a ser mucho mas cómodo testear asi con la lib que pasé en el otro PR

saantiaguilera replied:
Fixed in b11aa958aed8f3a5ac817b225dc757761272b54d

mastanca commented on src/api/service/user_service/user_service.go:
```diff
@@ -0,0 +1,30 @@
+
+var globalRef = &userService{}
+
+type userService struct {}
```
Siguiendo con la linea de arriba acá podrias hacer una interfaz y un "UserServiceImpl" que la implemente e inyectar el service en quien lo use

saantiaguilera replied:
Este no lo entendi. Ya estoy haciendo eso justamente.

`userService` implementa la interfaz [`service.UserService`](https://github.com/arpb2/C-3PO/blob/83dff1d2c2e52ee29c87dcdeda66754a902bb524/src/api/service/user_service.go#L5). Cada consumidor elije que implementacion usar de `service.UserService` (en los tests por ejemplo son mocks), y aca proveo una "impl" (este `userService` que ves) para que en produ se inyecte por ejemplo [aca](https://github.com/arpb2/C-3PO/blob/83dff1d2c2e52ee29c87dcdeda66754a902bb524/src/api/controller/code/code_get_controller.go#L25) (hay mas usos obvio).

Como no quiero tener muchas instancias del service (mas que nada para evitar tener muchas conexiones abiertas a una db/os/lo_que_sea, las obtengo via un [`GetService()`](https://github.com/arpb2/C-3PO/blob/83dff1d2c2e52ee29c87dcdeda66754a902bb524/src/api/service/user_service/user_service.go#L8) y no un `Create()`. De ahi que guardo el `globalRef` ese privado. Es analogo a un singleton, o eso intente simular.

Queria para todos los services tratar sus instancias asi. Obvio no serian vistos a traves de las imples (mucho menos como singleton accessors) sino siempre como sus interfaces, que bueno, por abajo terminan _o casualidad_ siendo eso.

Si es otra cosa que no estoy viendo, metemos una rapi call o audio por wpp y explicame!

mastanca replied:
Está bien, no habia entendido la idea del singleton. Si inyectas todo cuando creas la app creo que no hace falta tener la globalRef. Ponele cuando creas el server creas los controllers y les inyectas el service que le inyectas el dao. Ejemplo:

```go
	myDatastore := datastore.DB{}
	userService := service.UserServiceImpl{UserDatastore: myDatastore}
	userController := controller.UserControllerImpl{
		UserService: &userService,
	}
	loginController := controller.LoginControllerImpl{
		UserService: &userService,
	}
```

Con eso te ahorras tener la globalRef y es el mismo service. Si necesitas el mismo service en 2 lados (caso medio raro para nosotros pero ponele) tendrias que crearlo ahi. Fijate si te cierra, igual es lo mismo que estas haciendo vos pero de otra forma nomas. No mas preguntas señor juez

saantiaguilera replied:
Claro, es lo mismo solo que mi inyeccion inicial es "descentralizada" (cada controller de un negocio se crea desde un binder, pero los binders onda de `users` y `code` no se conocen entre si, asi que no pueden "compartir" services ni nada).

Talvez en alguna abstraccion de la abstraccion de la abstraccion lo refactorizo

mastanca reviewed (commented):
Un par de sugerencias

saantiaguilera committed:
[C3PO-13] Use testify/mock

- Use [testify/mock](github.com/stretchr/testify/mock) for mocking
objects in tests instead of creating mulitple implementations for each

saantiaguilera committed:
[C3PO-14] Use mock for auth_handler tests

- Start using stretchr mocks for authentication_handler tests
